Migrate parameter file + expand explanations (#356)
lethe-cfd/lethe: doc/source/parameters/cfd/non-linear_solver_control.rst

Non-linear Solver Control
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The Navier-Stokes equations (and other) are non-linear equations. The parameters in ``subsection non-linear solver``, whose default values are given in the text block below, control the non-linear solver used within Lethe.

.. code-block:: text

	subsection non-linear solver

> 	  # Non-linear solver that will be used. Choices is newton.

Reviewer: ```suggestion
	  # Non-linear solver that will be used. Choices are 
	  # <newton|inexact_newton|kinsol_newton>. 
```

	  set solver             = newton

> 	  # State whether from the non-linear solver should be printed Choices are
> 	  # <quiet|verbose>.
> 	  set verbosity          = verbose

Reviewer: ```suggestion
	  # State whether information from the non-linear solver should be printed.
	  # Choices are <quiet|verbose>.
	  set verbosity          = verbose
```

	  # Newton solver tolerance
	  set tolerance          = 1e-6

	  # Tolerance for the acceptation of the step
	  set step tolerance     = 0.9

	  # Maximum number of Newton Iterations
	  set max iterations     = 10
	end

> * The ``solver`` parameter enables to choose the nonlinear solver used. Currently, the only solver implemented is ``newton``, a Newton-Raphson solver which recalculates the jacobian matrix at every iteration (see the Theory Documentation).

Reviewer 2: There are some missing parameters (my fault). You can find them in the parameter template generated by the template executable. If you need help i'll write the text for them.

> * The ``solver`` parameter enables to choose the nonlinear solver used. Currently, the only solver implemented is ``newton``, a Newton-Raphson solver which recalculates the jacobian matrix at every iteration (see the Theory Documentation).

Author: No problem I'll do a first version so that you can expand on it :)

> * The ``solver`` parameter enables to choose the nonlinear solver used. Currently, Lethe supports two non-linear solver. The ``newton`` solver is a Newton-Raphson solver which recalculates the jacobian matrix at every iteration (see the Theory Documentation). The ``inexact_newton`` solver is a Newton-Raphson solver where the Jacobian matrix is reused between iterations. If the residual at iteration n+1 (:math:`R_{n+1}`) newton iteration is  lower than ``matrix_tolerance`` times the residual at iteration n, the Newton iteration will keep using the same jacobian matrix. Setting ``reuse matrix= true`` enables the usage of the same jacobian matrix for the following non-linear problem. This approach can be worthwhile in transient simulations with a small time-step. The goal is to seek a compromise between the cost of assembling the matrix and the preconditioner versus the cost of solving the linear system of equations.

Reviewer: ```suggestion
* The ``solver`` parameter enables choosing the nonlinear solver for the simulation. Currently, Lethe supports three non-linear solvers. The ``newton`` solver is a Newton-Raphson solver which recalculates the Jacobian matrix at every iteration (see the Theory Documentation). The ``inexact_newton`` solver is a Newton-Raphson solver where the Jacobian matrix is reused between iterations. If the residual at Newton iteration n+1 (:math:`R_{n+1}`) is  lower than ``matrix_tolerance`` times the residual at iteration n, the Newton iteration will keep using the same Jacobian matrix. Setting ``reuse matrix= true`` enables the usage of the same Jacobian matrix for the following non-linear problem. This approach can be worthwhile in transient simulations with a small time-step. The goal is to seek a compromise between the cost of assembling the matrix and the preconditioner versus the cost of solving the linear system of equations. The third option is an interface that uses through deal.II the Newton-Raphson solver as implemented in the `Sundials library <https://computing.llnl.gov/projects/sundials/kinsol>`_. This non-linear solver is still being tested.

```

Author: thank you I was looking for that!

> * The ``verbosity`` options enables to display the residual at each non-linear iteration, to monitor how the non-linear iterations are progressing.

Reviewer: ```suggestion
* The ``verbosity`` option enables the display of the residual at each non-linear iteration, to monitor how the non-linear iterations are progressing.
```

Author: Thank you, I correct these typo on the next pull :)

.. note::
	The residual should decrease rapidly between Newton iterations.

> 	Exemple of a ``set verbosity = verbose`` output:

Reviewer: ```suggestion
	Example of a ``set verbosity = verbose`` output:
```

	
	.. code-block:: text

		Newton iteration: 0  - Residual:  429.541
				alpha = 1      res = 0.3705
		Newton iteration: 1  - Residual:  0.3705
				alpha = 1      res = 0.0001044
		Newton iteration: 2  - Residual:  0.0001044
				alpha = 1      res = 1.12e-08
		Newton iteration: 3  - Residual:  1.12e-08
				alpha = 1      res = 1.547e-12

* The ``step tolerance`` parameter controls how much the L2 norm of the residual must decrease to proceed to the next non-linear step. If the ``new_residual``:math:`<` ``old_residual``:math:`\times` ``step tolerance``, then a Newton iteration is accepted. If this condition is not reached, then a relaxation of the step is applied (increasing the ``alpha`` parameter, as printed on the terminal if ``set verbosity = verbose``) until this condition is reached.
* The ``tolerance`` parameter controls the value under which the residual must be to proceed to the next iteration.
.. hint::
	The ``tolerance`` parameter is directly linked to the numerical convergence of the simulation, but also to the computational cost (number of Newton iteration).

> 	For simple simulations, the step tolerance can be quite low, for instance ``set tolerance = 1e-12``. However, such a tolerance can be impossible to attain for more complex simulations : the step tolerance of the non-linear solver can be increased, for instance ``set tolerance = 1e-4``

Reviewer 2: ```suggestion
	For simple simulations, the tolerance can be set quite low, for instance ``set tolerance = 1e-12``. However, such a tolerance can be impossible to attain for more complex simulations : the step tolerance of the non-linear solver can be increased, for instance ``set tolerance = 1e-4``
```

> * The ``max iterations`` parameter sets a hard limit to the number of Newton iterations, even if the ``tolerance`` is not reached.

Reviewer 2: ```suggestion
* The ``max iterations`` parameter sets a hard limit to the number of Newton iterations, even if the ``tolerance`` is not reached.
```

.. warning::
   Be careful to always set an absolute tolerance for the linear solver that is below the tolerance of the non-linear 
   solver.  Otherwise, you might find that it is impossible to converge because the linear system of equation is solved with insufficient accuracy.
